man/tpm2_decapsulate.1.md
% tpm2_decapsulate(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_decapsulate**(1) - Perform a KEM decapsulation with a private TPM key.

# SYNOPSIS

**tpm2_decapsulate** [*OPTIONS*]

# DESCRIPTION

**tpm2_decapsulate**(1) - This command performs the private key operation of a 
Key Encapsulation Mechanism (KEM) using a TPM private key.

Given a private key and a ciphertext produced by an earlier encapsulation operation
such as **TPM2_Encapsulate()**, the command returns the same **sharedSecret** that
was produced during encapsulation. 


# OPTIONS

  * **-c**, **\--object-context**=_OBJECT_:

    Reference to the loaded KEM key to use for decapsulation.

  * **-p**, **\--auth**=_AUTH_:

    The authorization value for the key.

  * **-i**, **\--ciphertext**=_FILE_:
  
	Input file path containing the encapsulated ciphertext.
	
  * **-s**, **\--shared-secret**=_FILE_:
  
	Output file path for the decaspsulated shared secret


## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

## Setup
The first step is to create the primary object with mlkem.

```bash
tpm2 createprimary -C o -G mlkem1024 -g sha256 -c mlkem.ctx -a 'fixedtpm|fixedparent|sensitivedataorigin|userwithauth|decrypt'
```

Step 2 is to encapsulate.

```bash
tpm2_encapsulate -c mlkem.ctx --shared-secret=secret.bin --ciphertext=ciphertext.bin
```

Step 3 is to decapsulate.

```bash
tpm2_decapsulate -c mlkem.ctx --ciphertext=ciphertext.bin --shared-secret=secret_dec.bin
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_decapsulate.c
/* SPDX-License-Identifier: BSD-3-Clause */
/***********************************************************************;
* Copyright (c) 2026, STMicroelectronics
*
* All rights reserved.
***********************************************************************/

#include <errno.h>
#include <inttypes.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_capability.h"
#include "tpm2_tool.h"
#include "tpm2_alg_util.h"
#include "tpm2_util.h"

typedef struct tpm_decapsulate_ctx tpm_decapsulate_ctx;
struct tpm_decapsulate_ctx {
	
    char *context_arg;
    char *auth_arg;
    char *in_ciphertext_file;
    char *out_secret_file;
    tpm2_loaded_object context_object;
};

static tpm_decapsulate_ctx ctx = {0};

static bool load_ciphertext(const char *path, TPM2B_KEM_CIPHERTEXT *ciphertext){
	
	FILE *f = fopen(path, "rb");
	
	if(!f){
		LOG_ERR("Cannot open ciphertext file : %s", path);
		return false;
	}
	
	size_t n = fread(ciphertext->buffer, 1, sizeof(ciphertext->buffer), f);
	if (ferror(f)){
		fclose(f);
		LOG_ERR("Error while reading ciphertext file: %s", path);
		return false;
	}
	
	fclose(f);
	
	if (n > sizeof(ciphertext->buffer)){
		LOG_ERR("Ciphertext file too large");
		return false;
	}
	
	ciphertext->size = (UINT16)n;
	return true;
	
}

static tool_rc check_options(void){
	
	if(!ctx.context_arg){
		LOG_ERR("Expected option -c for object context");
		return tool_rc_option_error;
	}
	
	if(!ctx.in_ciphertext_file){
		LOG_ERR("Expected option -i for ciphertext input");
		return tool_rc_option_error;
	}
	
	if(!ctx.out_secret_file){
		LOG_ERR("Expected option -s for shared secret output");
		return tool_rc_option_error;
	}
	
	return tool_rc_success;
}
	

static tool_rc decapsulate_and_save(ESYS_CONTEXT *ectx) {

    TPM2B_SHARED_SECRET *sharedSecret = NULL;
    TPM2B_KEM_CIPHERTEXT ciphertext = {0};

    bool ret = load_ciphertext(ctx.in_ciphertext_file, &ciphertext);
    if (!ret) {
		return tool_rc_general_error;
	}
	
	TSS2_RC rval = Esys_Decapsulate(ectx,
									ctx.context_object.tr_handle,
									ESYS_TR_PASSWORD,
									ESYS_TR_NONE,
									ESYS_TR_NONE,
									&ciphertext,
									&sharedSecret);
	
	if (rval != TSS2_RC_SUCCESS){
		LOG_PERR(Esys_Decapsulate, rval);
		return tool_rc_general_error;
	}
	
	if (!sharedSecret){
		LOG_ERR("TPM returned NULL shared secret");
		return tool_rc_general_error;
	}
	
	tpm2_tool_output("shared secret: ");
	for (UINT16 i=0; i < sharedSecret->size; i++){
		tpm2_tool_output("%02x", sharedSecret->buffer[i]);
	}
	tpm2_tool_output("\n");
	
	ret = files_save_bytes_to_file(ctx.out_secret_file,
								   sharedSecret->buffer,
								   sharedSecret->size);
	if (!ret) {
		LOG_ERR("Cannot save shared secret");
		Esys_Free(sharedSecret);
		return tool_rc_general_error;
	}
    Esys_Free(sharedSecret);
    return tool_rc_success;        
    
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'c':
        ctx.context_arg = value;
        break;
    case 's':
        ctx.out_secret_file = value;
        break;
    case 'p':
        ctx.auth_arg = value;
        break;
    case 'i' : 
		ctx.in_ciphertext_file = value;
		break;
    default:
		return false;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
        { "object-context",    required_argument, NULL, 'c' },
        { "shared-secret",     required_argument, NULL, 's' },
        { "auth",              required_argument, NULL, 'p' },
        { "ciphertext",        required_argument, NULL, 'i' }
    };

    *opts = tpm2_options_new("c:p:i:s:", ARRAY_LEN(topts), topts, on_option,
            NULL, 0);

    return *opts != NULL;
}

static tool_rc init(ESYS_CONTEXT *ectx) {

    tool_rc rc = tpm2_util_object_load_auth(ectx, ctx.context_arg,
		ctx.auth_arg, &ctx.context_object, false,
		TPM2_HANDLE_ALL_W_NV);
    if (rc != tool_rc_success) {
        return rc;
    }

    return tool_rc_success;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *context, tpm2_option_flags flags) {
    
    UNUSED(flags);
    
    tool_rc rc = check_options();
    if (rc != tool_rc_success){
		return rc;
	}

    rc = init(context);
    if (rc != tool_rc_success) {
        return rc;
    }

    return decapsulate_and_save(context);
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("decapsulate", tpm2_tool_onstart, tpm2_tool_onrun, NULL, NULL)
